add specific exception for time jumping backwards

jowave opened the pull request:
reason: ROS shutdown and time jumps should be distinguishable.
ROSTimeMovedBackwardsException inherits from ROSInterruptException
to preserve the API.
fixes #485

jowave commented:
Recreated the pull request against indigo-devel.
Sorry for the inconvenience.

ros-pull-request-builder commented:
Can one of the admins verify this patch?

dirk-thomas commented:
@ros-pull-request-builder

dirk-thomas commented:
Can you please move the exception class to `clients/rospy/src/rospy/exceptions.py` where most other exceptions are defined.

Beside this minor detail the patch lgtm.

jowave commented:
I moved the exception and updated the pull request.

dirk-thomas commented:
@ros-pull-request-builder

ros-pull-request-builder commented:
Test failed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/121/

dirk-thomas commented:
Please update your patch to pass all unit tests. It currently breaks 145 tests.

dirk-thomas commented:
Please update your patch to pass all unit tests. It currently breaks 145 tests.

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/131/

jowave commented:
I have updated the patch. Apparently it passed the tests now.

wjwwood commented:
+1

dirk-thomas commented on clients/rospy/src/rospy/timer.py:
```diff
@@ -128,7 +131,8 @@ def sleep(duration):
 
         if rospy.rostime.get_rostime() < initial_rostime:
-            raise rospy.exceptions.ROSInterruptException("ROS time moved backwards")
+            rospy.core.logerr("ROS time moved backwards")
```
Can you please include the time offset in the error message so that the user has a better idea how much the time jumped.

jowave commented:
I added the time offset in the error message and a small test for the exception.

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/134/

tlind commented:
I am worried that commit d934e53 might break existing code which is trying to catch the exception. Since there was previously no specific exception for this case, the only way of specifically catching the time-moved-backwards issue was to check the exception message, e.g.:

```
try:
    r.sleep()
except rospy.exceptions.ROSInterruptException as e:
    if e.message == "ROS time moved backwards":
        rospy.logwarn("Time has moved backwards (probably ROS clock has been reset)!")
```

However, it seems that supplying the time offset as an additional argument to the exception constructor will break that message:

```
>>> print rospy.exceptions.ROSInterruptException("ROS time moved backwards").message
ROS time moved backwards
>>> print rospy.exceptions.ROSInterruptException("ROS time moved backwards", -10.0).message
[empty string]
```

Might it be a better idea to only rospy.logerr() the time offset, and leave the exception message as-is? Or add an additional field to the exception class which contains the time offset?

EDIT: I just read that Exception.message seems to be deprecated since 2.6, though I've been using it a few times and never saw a warning anywhere. Not sure how much other code depends on it. The correct way seems to be to use e.args[0], which returns the same result in both cases.

dirk-thomas commented:
@tlind That is a good point.

@jowave Can you please update the new exception class to store the time offset in a member variable and keep the message of the exception as-is.

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/135/

jowave commented:
I modified the code to work with Exception.message and added a test to check for the right message.

dirk-thomas commented on test/test_rospy/test/unit/test_rospy_exceptions.py:
```diff
@@ -62,3 +62,17 @@ def test_ROSInterruptException(self):
+        except ROSInterruptException as e:
+            # ensure the message is not changed, because old code may check it
+            self.assertEqual("ROS time moved backwards", e.message)
+            pass
```
This line should be obsolete.

dirk-thomas commented:
Please squash your commits after removing the unnecessary `pass` line.

Commit:
add specific exception for time jumping backwards

reason: ROS shutdown and time jumps should be distinguishable.
ROSTimeMovedBackwardsException inherits from ROSInterruptException
to preserve the API.
fixes #485

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/138/

dirk-thomas commented:
Thank you for you efforts on this. Merging...

jowave commented:
You're welcome. Thank you too.

Edit: Hmm, the ros-pull-request-builder is answering. Should i not have replied to a closed pull request?

dirk-thomas commented:
No worries. The Jenkins check is still experimental. We hopefully can improve that in the future.

peci1 commented on clients/rospy/src/rospy/timer.py:
```diff
@@ -128,7 +131,9 @@ def sleep(duration):
         if rospy.rostime.get_rostime() < initial_rostime:
-            raise rospy.exceptions.ROSInterruptException("ROS time moved backwards")
+            time_jump = (initial_rostime - rospy.rostime.get_rostime()).to_sec()
+            rospy.core.logerr("ROS time moved backwards: %ss", time_jump)
```
Is it really needed to log the exception? I use ROS nodes together with Gazebo, where restarting the time is quite normal, and these messages start being annoying. The previous behavior also did not log the exception.

dirk-thomas replied:
I agree with you that the log message is not strictly necessary. On the other hand this code has been integrated over two years ago and since then been released in three ROS distributions. Since no concerns were raised in the meantime I don't see the log message as a problem either.

Please feel free to create a pull request against the latest development branch (currently *lunar-devel*).
